libp2p/connmanager.nim

```diff
@@ -331,35 +345,120 @@ proc storeConn*(c: ConnManager, conn: Connection) =
 
   c.conns[peerId].incl(conn)
-
-  # Launch on close listener
-  # All the errors are handled inside `onClose()` procedure.
-  asyncSpawn c.onClose(conn)
   libp2p_peers.set(c.conns.len.int64)
 
   trace "Stored connection",
     conn, direction = $conn.dir, connections = c.conns.len
 
+proc trackConn(c: ConnManager,
+               provider: ConnProvider,
+               dir: Direction):
+               Future[Connection] {.async.} =
+  var conn: Connection
+  try:
+    conn = await provider()
+
+    if isNil(conn):
+      return
+
+    conn.dir = dir
+    trace "Got connection", conn, dir = $dir
+
+    proc semaphoreMonitor() {.async.} =
+      try:
+        await conn.join()
+      except CatchableError as exc:
+        trace "Exception in semaphore monitor, ignoring", exc = exc.msg
+
+      if conn.dir == Direction.In:
+        c.inConnSemaphore.release()
+      else:
+        c.outConnSemaphore.release()
+
+    asyncSpawn semaphoreMonitor()
+  except CatchableError as exc:
+    trace "Exception tracking connection", exc = exc.msg
+    if not isNil(conn):
+      await conn.close()
+
+    raise exc
+
+  return conn
+
+proc trackIncomingConn*(c: ConnManager,
+                        provider: ConnProvider):
+                        Future[Connection] {.async.} =
+  ## await for a connection slot before attempting
+  ## to call the connection provider
+  ##
+
+  var conn: Connection
+  try:
+    trace "Tracking incoming connection"
+    await c.inConnSemaphore.acquire()
+    conn = await c.trackConn(provider, Direction.In)
+    if isNil(conn):
+      trace "Couldn't acquire connection, releasing semaphore slot", dir = $Direction.In
+      c.inConnSemaphore.release()
+  except CatchableError as exc:
+    trace "Exception tracking connection", exc = exc.msg
+    c.inConnSemaphore.release()
+    raise exc
+
+  return conn
+
+proc trackOutgoingConn*(c: ConnManager,
+                        provider: ConnProvider):
+                        Future[Connection] {.async.} =
+  ## try acquiring a connection if all slots
+  ## are already taken, raise TooManyConnectionsError
+  ## exception
+  ##
+
+  trace "Tracking outgoing connection", count = c.outConnSemaphore.count,
+                                        max = c.outConnSemaphore.size
+
+  if not c.outConnSemaphore.tryAcquire():
+    trace "Too many outgoing connections!", count = c.outConnSemaphore.count,
+                                            max = c.outConnSemaphore.size
+    raise newTooManyConnectionsError()
+
+  var conn: Connection
+  try:
+    conn = await c.trackConn(provider, Direction.Out)
+    if isNil(conn):
+      trace "Couldn't acquire connection, releasing semaphore slot", dir = $Direction.Out
+      c.outConnSemaphore.release()
+  except CatchableError as exc:
+    trace "Exception tracking connection", exc = exc.msg
+    c.outConnSemaphore.release()
+    raise exc
+
+  return conn
+
 proc storeOutgoing*(c: ConnManager, conn: Connection) =
   conn.dir = Direction.Out
```

Comment:
isn't the connection direction already set correctly when the connection instance is created? it would be better if it was immutable from that point onwards

Reply:
Yes, you have a point. It's probably a good idea to add an assert so that only the correct direction is added.

Reply:
or just have a `storeconnection`  which reads `dir` from `conn`

Reply:
I think the assert is safer, in case we inadvertently set the wrong direction somewhere?

Reply:
well, it says the same thing in a different way - if `dir` can't be relied upon, there will be lots of other issues as well - fewer store functions seems more simple, and whoever is calling storeoutgoing right now can potentially do the assert, if it makes sense in that context - generally, these kinds of issues appear usually with mutable data.

Reply:
I'm just trying to think of a convenient API here, this aren't anything but helpers and they are meant to do just that, removing them will not significantly affect anything. The assert is there just to make sure that the direction hasn't been changed inadvertently, which can also be accomplished by making it read-only.